client/app/components/Select.tsx

```diff
@@ -0,0 +1,40 @@
+import React, { useRef, useState, useEffect } from "react";
+import { sortBy, get } from "lodash";
+import AntdSelect, { SelectProps } from "antd/lib/select";
+
+function getItemOfPercentile<T>(list: Array<T>, percentile: number, sortIteratee = "length") {
+  if (get(list[0], sortIteratee, null) === null) {
+    return;
+  }
+  const sortedList = sortBy(list, sortIteratee);
+  const percentileIndex = Math.ceil((list.length - 1) * (percentile / 100));
+  return sortedList[percentileIndex];
+}
+
+const FONT_RATIO = 0.7;
+
+function Select({ style, options, ...props }: SelectProps<any>): JSX.Element {
+  const [dropdownMatchSelectWidth, setDropdownMatchSelectWidth] = useState<number | boolean>(false);
+  useEffect(() => {
+    if (options && options.length > 400) {
+      // using body because it's too messy to get a reference to the element itself
+      const fontSize = parseFloat(getComputedStyle(document.body).fontSize);
+      const itemOf80thPercentile = getItemOfPercentile(options, 80, "label.length");
+      if (itemOf80thPercentile) {
+        const len = String(itemOf80thPercentile.label).length;
+        setDropdownMatchSelectWidth(len * fontSize * FONT_RATIO);
```

Comment:
There is a way to measure string more precisely: https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/measureText
Why: fontSize is a _height_ of the font, but you need a _width_. Of course in this case it could be a good enough approximation, but if there's a way to do things properly - why not?

Reply:
My first thought on avoiding that was that it could be slow for huge lists (>10k elements), but I didn't test it 🤔. If that works fast for, let's say, 100k strings, we can create a helper (similar to the `url` service) to make the right measurement.

Considering it accurate, we can actually use the actual Max from the list (not necessarily the 80th percentile anymore), bu that's at will.

Reply:
Actually, this measurement is needed only once, for the single item (am I right, @rafawendel?). So if could be very accurate (but I'm still not sure - should it be so accurate? 🤔)

Reply:
Another question that crossed my mind just now: if we have all items, and sort them by length, - why do we use 80th percentile item instead of longest one?

Reply:
> Actually, this measurement is needed only once, for the single item (am I right, @rafawendel?). So it could be very accurate (but I'm still not sure - should it be so accurate? 🤔)

Although very unlikely, since the width per char can differ, it should be possible (I'm actually not sure if this case exists) to have a string with more characters to end up with less width. But we can assume that this won't happen 😆.

> Another question that crossed my mind just now: if we have all items, and sort them by length, - why do we use 80th percentile item instead of longest one?

I think the idea was to not grow it unnecessarily by excluding too large items from the calculation. To me this made more sense because the calculation was not really accurate.

Reply:
Let's create this helper using a `<canvas>` (similar to `client/app/services/url.js`), this will be useful for not so simple cases, e.g.: To make a custom calculation when the Option is a Node (`dropdownMatchSelectWidth={iconWidth + anyMargin + textWithMaxLengthWidth}` for the Data Source dropdown)

Reply:
E.g. "www" (3 chars) is wider than "iiiii" (5 chars), so it's possible 🙂 Canvas.measureText will give us a proper text width in pixels.

![image](https://user-images.githubusercontent.com/12139186/96884710-7a032180-148a-11eb-8d28-a428e2c155c4.png)

Reply:
> E.g. "www" (3 chars) is wider than "iiiii" (5 chars), so it's possible 🙂 Canvas.measureText will give us a proper text width in pixels.
> 
> ![image](https://user-images.githubusercontent.com/12139186/96884710-7a032180-148a-11eb-8d28-a428e2c155c4.png)

Yes, but in this case we would calculate based on the "iiiii", not on the "www", because it has greater length

Reply:
> Another question that crossed my mind just now: if we have all items, and sort them by length, - why do we use 80th percentile item instead of longest one?

I actually opted for the percentile because I wanted the size not to be skewed by outliers, e.g. one single very large word that would cause the width to be really large. At the beginning I thought on using average + 2 standard deviations, but the implementation would be a little cumbersome, then I decided to approximate it with percentiles.

Surely a solution with averages would be only O(n), and sorting is O(n log n), which is worse, but I didn't sound like too much of a problem, as having too much of my mind on optimizing it would make me waste a lot of time trying to find ways to avoid re-renders or re-calculating on every render.